[py] Server class to manage (download/run) grid server

cgoldberg opened the pull request:
### **User description**
### 🔗 Related Issues

This is the Python implementation for #12305 

### 💥 What does this PR do?

This PR adds a new `selenium.webdriver.remote.server` module with a `Server` class. It is used for downloading/starting/stopping the grid server in standalone mode.

It uses Selenium Manager to download and locate the server jar file. To use the `start()` method, you must have a Java JRE on your system (or it will give you an error message).

This also updates the PyTest configuration (`py/conftest.py`) to use this class to replace the code for launching the server for remote tests. It uses the existing local server instead of downloading one, since it is already built locally and CI is setup to use it.

----

### Example usage:

```
from selenium import webdriver
from selenium.webdriver.remote.server import Server

server = Server()
server.start()
driver = webdriver.Remote(options=webdriver.ChromeOptions())
# do some webdrivery stuff
driver.quit()
server.stop()
```

### More examples:

- download the latest server and run it:
```
from selenium.webdriver.remote.server import Server

server = Server()
server.start()
server.stop()
```

- download a specific version of the server and run it:
```
from selenium.webdriver.remote.server import Server

server = Server(version="4.30.0")
server.start()
server.stop()
```

- download the latest server and run it on a specified ip/port:
```
from selenium.webdriver.remote.server import Server

server = Server(host="127.0.0.1", port=8888)
server.start()
server.stop()
```

- download the latest server and run it with an environment variable set:
```
import os
from selenium.webdriver.remote.server import Server

env = os.environ.copy()
env["MY_VARIABLE"] = "special"
server = Server(env=env)
server.start()
server.stop()
```

- run an existing server without downloading:
```
from selenium.webdriver.remote.server import Server
server = Server(path="/path/to/selenium-server-4.31.0.jar")
server.start()
server.stop()
```

### 💡 Additional Considerations

I added some unit tests, but they aren't comprehensive because it's a pain to mock selenium manager and the subprocess call... I'll figure that out later.

### 🔄 Types of changes
<!-- ✂️ Please delete anything that doesn't apply -->
- New feature (non-breaking change which adds functionality *and tests!*)


___

### **PR Type**
Enhancement, Tests


___

### **Description**
- Introduces `Server` class for managing Selenium Grid server.
  - Supports downloading, starting, and stopping the server.
  - Allows specifying host, port, version, path, and environment.

- Refactors pytest fixture to use new `Server` class for remote tests.

- Adds unit tests for `Server` class error handling and validation.

- Updates API documentation to include new `server` module.


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>server.py</strong><dd><code>Add Server class for Selenium Grid server management</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/selenium/webdriver/remote/server.py

<li>Adds new <code>Server</code> class for managing Selenium Grid server lifecycle.<br> <li> Implements validation for path, port, and version.<br> <li> Handles server startup, shutdown, and error scenarios.<br> <li> Integrates with Selenium Manager for automatic server download.


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15666/files#diff-f402a040be51e4850063837cff55fc5b6b6817bf4c242c37e51ad8b2dcc02b3b">+142/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>conftest.py</strong><dd><code>Refactor pytest server fixture to use Server class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/conftest.py

<li>Refactors remote server pytest fixture to use new <code>Server</code> class.<br> <li> Removes legacy subprocess/socket logic for server startup.<br> <li> Ensures environment variables are handled for Linux compatibility.


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15666/files#diff-98e9e290ede5d0c7ac9e7df5b49edf63f5df5fcc946b7736e4a6139e4cda26e3">+13/-55</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>remote_server_tests.py</strong><dd><code>Add unit tests for Server class validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/test/unit/selenium/webdriver/remote/remote_server_tests.py

<li>Adds unit tests for <code>Server</code> class input validation and error handling.<br> <li> Tests invalid path, version, port, and stopping server not running.


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15666/files#diff-778bf38c1d07a3bffb033ebe6e9b546b7a14e1dc02f7541d11e72e8a515dd845">+52/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>api.rst</strong><dd><code>Document remote.server module in API docs</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/docs/source/api.rst

<li>Documents new <code>selenium.webdriver.remote.server</code> module in API <br>reference.


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15666/files#diff-7e8f0a6376548281d6c1a5b57121f0a701a76c5f88129fe7def3d89f39b63c43">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

cgoldberg committed:
[py] Add class to manage remote/grid server

cgoldberg committed:
[py] Add class to manage remote/grid server

cgoldberg committed:
[py] User Server class in conftest

cgoldberg committed:
[py] Add server tests

cgoldberg committed:
Merge branch 'SeleniumHQ:trunk' into py-server-standalone-control

cgoldberg committed:
[py] Add another test

cgoldberg committed:
[py] Add exception

cgoldberg committed:
[py] Change exception type

cgoldberg committed:
[py] Check if process is running before terminating

cgoldberg committed:
[py] Close socket

cgoldberg committed:
[py] Update docstrings

cgoldberg committed:
Merge branch 'SeleniumHQ:trunk' into py-server-standalone-control

cgoldberg committed:
[py] Fix status_url

shbenzer reviewed (dismissed):
LGTM. Thank you @cgoldberg!

shbenzer commented:
Does this pr need a similar change:

```ruby
        def grid_path(required_version)
          command = [binary, '--grid']
          command << required_version if required_version

          location = run(*command)
          WebDriver.logger.debug("Grid found at #{location}", id: :selenium_manager)
          Platform.assert_file location

          location
        end

        private

        def generate_command(binary, options)
          command = [binary, '--browser', options.browser_name, '--output', 'json']
          command = [binary, '--browser', options.browser_name]
          if options.browser_version
            command << '--browser-version'
            command << options.browser_version
            
@@ -65,7 +76,6 @@ def generate_command(binary, options)
            command << '--proxy'
            (command << options.proxy.ssl) || options.proxy.http
          end
          command << '--debug' if WebDriver.logger.debug?
          command
        end

@@ -98,6 +108,8 @@ def binary
        end

        def run(*command)
          command += %w[--output json]
          command << '--debug' if WebDriver.logger.debug?
          WebDriver.logger.debug("Executing Process #{command}", id: :selenium_manager)

          begin
```

from `rb/lib/selenium/webdriver/common/selenium_manager.rb` from #12299 

to be fully functional? Or is the functionality to download the grid .jar file already built into Selenium Manager w/ the --grid command?

Similarly, since if I'm not mistaken, the CI has the .jar file included, should we add a test for this download functionality?

cgoldberg commented:
> Or is the functionality to download the grid .jar file already built into Selenium Manager w/ the --grid command?

Yes, exactly... with this:
```
args = ["--grid"]
if self.version:
    args.append(self.version)
self.path = selenium_manager.binary_paths(args)["driver_path"]
```

Calling that with the `--grid` arg will check if the .jar exists in the default location, and download it if it doesn't. You can add a version number to get a specific version. This functionality is skipped if you create the `Server()` with the `path` argument. It will then use the .jar in the location you specified and not download anything.

>  the CI has the .jar file included

CI already has the jar, so it is configured to just use that one.

> should we add a test for this download functionality?

Probably. I was going to add some unit tests with everything mocked, so it wouldn't actually download, just make sure everything was called correctly. But I can add an integration test where it actually downloads the .jar and verifies it on disk.

shbenzer commented:
Perfect, great work!

shbenzer commented:
Ah, I see why I was confused - I thought `rb/lib/selenium/webdriver/common/selenium_manager.rb` was part of the Selenium Manager's code, but it's written in rust not ruby.

titusfortner commented:
The Ruby version of this implements more options. Not sure if they are all necessary, I will say that being able to toggle the logging level for tests is nice when debugging.

https://github.com/SeleniumHQ/selenium/blob/trunk/rb/lib/selenium/server.rb#L183

cgoldberg committed:
[py] Move download to nmew method

cgoldberg commented:
I added a new `download_if_needed()` method to make it more testable, and added a test for that.

There are 2 minor concerns:

- The test downloads the .jar, but the test fixture also downloads the .jar, so it gets downloaded twice. This is necessary because the fixture deletes the .jar before and after the test runs, so it has clean state for the next run. I don't see any way around this, because there doesn't seem to be an easy way to get the .jar file name from Selenium Manager without actually downloading the file.

- I can't find a way to get Selenium Manager to download to any other location. This isnt a big deal for CI or when running the other tests locally. But it will cleanup/delete the downloaded installation when this tests finishes, so that might be annoying if you had downloaded it previously.

I don't think either of these are a big deal, but thought I would point them out.

cgoldberg commented:
> I will say that being able to toggle the logging level for tests is nice when debugging.

I'll add `--log-level`, but I'm going to hold off on adding all of the options for this first version. If anyone actually uses this, we can add more. The server itself accepts like 5000 options :)

titusfortner commented:
Also for reference, the Ruby tests start the server like this:
https://github.com/SeleniumHQ/selenium/blob/trunk/rb/spec/integration/selenium/webdriver/spec_support/test_environment.rb#L90
and get the server like this:
https://github.com/SeleniumHQ/selenium/blob/trunk/rb/spec/integration/selenium/webdriver/spec_support/test_environment.rb#L132

The idea is that if you're running without bazel it will download it, and running with bazel it will use what bazel built
Not sure it matters as much now that everything is mostly working well with bazel

cgoldberg committed:
[py] Add log_level arg

cgoldberg commented:
I added the `log_level` arg when creating the Server. It accepts only valid log levels: "SEVERE", "WARNING", "INFO", "CONFIG", "FINE", "FINER", "FINEST".

cgoldberg committed:
[py] Fix f-string

cgoldberg commented:
> The idea is that if you're running without bazel it will download it, and running with bazel it will use what bazel built

That sounds better. The way it works in the Python tests is that it always looks for what bazel built.. so if you don't build grid with bazel first, it will just fail.

I'll add that to conftest.py

cgoldberg committed:
[py] Fix instance check and add unit tests

cgoldberg committed:
[py] Update log message

cgoldberg committed:
[py] Download server if none exists from bazel build

cgoldberg committed:
[py] Fix print

cgoldberg committed:
[py] Fix linting error

cgoldberg committed:
Merge branch 'SeleniumHQ:trunk' into py-server-standalone-control

cgoldberg committed:
Merge branch 'SeleniumHQ:trunk' into py-server-standalone-control

cgoldberg committed:
Merge branch 'SeleniumHQ:trunk' into py-server-standalone-control
